Fixes #32860 - move toast list to app's root

Ron-Lavi opened the pull request:
Toasts were triggered only from pages that used the PageLayout component,
but now they will be triggered directly from the root of the app.


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

amirfefer commented on webpack/assets/javascripts/react_app/Root/ReactApp.js:
```diff
@@ -19,7 +20,10 @@ const ReactApp = ({ layout, metadata, toasts }) => {
           <ConnectedRouter history={history}>
             <Layout data={layout}>
-              <AppSwitcher />
+              <>
+                <ToastsList railsMessages={toasts} />
+                <AppSwitcher />
+              </>
```
Why do we need a fragment wrapper here?
```suggestion
<Layout data={layout}>
   <ToastsList toasts={toasts} />
   <AppSwitcher />
</Layout>
            
```

Ron-Lavi replied:
thanks, no need :)

amirfefer commented on webpack/assets/javascripts/react_app/Root/ReactApp.js:
```diff
@@ -8,6 +8,7 @@ import Layout, { propTypes as LayoutPropTypes } from '../components/Layout';
+import ToastsList from '../components/ToastsList';
 
 const ReactApp = ({ layout, metadata, toasts }) => {
   const contextData = { metadata, toasts };
```
Do we still need `toasts` in the foreman context?

Ron-Lavi replied:
I was also thinking about removing it, though I was too afraid to break plugins...
I guess we could first remove the usage of useForemanContext with plugins and then remove it as well.

amirfefer replied:
How about a deprecation warning?

Ron-Lavi replied:
Seems that there's no helper method specific to toasts, so it would be kinda hard to deprecate,
This is what we have in `ForemanContext.js`

```js
export const useForemanContext = () => React.useContext(getForemanContext());

const useForemanMetadata = () => useForemanContext().metadata;

export const useForemanVersion = () => useForemanMetadata().version;
export const useForemanSettings = () => useForemanMetadata().UISettings;
export const useForemanDocUrl = () => useForemanMetadata().docUrl;
export const useForemanOrganization = () => useForemanMetadata().organization;
export const useForemanLocation = () => useForemanMetadata().location;
export const useForemanUser = () => useForemanMetadata().user;
```

Added a subtask of the issue I opened in Redmine: https://projects.theforeman.org/issues/32871
so we could better tackle this across plugins.

amirfefer replied:
Looks like just [webhooks plugin](https://github.com/theforeman/foreman_webhooks/blob/1aa674a4e6dd37a37b2e93e2d77d4a6ffb8cb071/webpack/ForemanWebhooks/Routes/Webhooks/WebhooksIndexPage/index.js#L42) consumes toasts from `ForemanContext`, worth a quick fix

Ron-Lavi replied:
nice catch @amirfefer! opened https://github.com/theforeman/foreman_webhooks/pull/39

tbrisker replied:
We can drop it from the context in a follow-up pr, right now this shouldn't be a breaking change afaict.

tbrisker replied:
actually, if we don't include it in the context (or maybe include an empty array instead of the actual toasts), the webhooks plugin should function just fine until it's cleaned up as it won't be displaying any toasts on its own and we won't get duplicates

amirfefer reviewed (changes requested):
Thanks @laviro 👍

Ron-Lavi commented:
Ready for another round @amirfefer :)

MariaAga commented:
Currently notifications are shown twice, should `ToastList` be removed from:
https://github.com/theforeman/foreman/blob/330d63481f23bf1304eda460dcd7d9acc9d1492d/app/helpers/application_helper.rb#L479
https://github.com/theforeman/foreman/blob/330d63481f23bf1304eda460dcd7d9acc9d1492d/app/views/layouts/_application_content.html.erb#L3
?
Tested on user create\delete notifications

Ron-Lavi commented on app/views/layouts/login.html.erb:
```diff
@@ -1,7 +1,6 @@
 <% title(_("Login")) %>
 
 <%= content_for(:content) do %>
-  <%= notifications %>
```
alerts are already passed in `login_helper.rb`

Ron-Lavi commented:
> Currently notifications are shown twice, should `ToastList` be removed from:
> https://github.com/theforeman/foreman/blob/330d63481f23bf1304eda460dcd7d9acc9d1492d/app/helpers/application_helper.rb#L479
> 
> 
> https://github.com/theforeman/foreman/blob/330d63481f23bf1304eda460dcd7d9acc9d1492d/app/views/layouts/_application_content.html.erb#L3
> 
> 
> ?
> Tested on user create\delete notifications

Thanks @MariaAga, fixed, can you test it again?

tbrisker reviewed (dismissed):
Nice cleanup @laviro !

amirfefer reviewed (commented):
Some plugins fail with `undefined local variable or method notifications for` because of dropping 
```rb
#application_helper.rb
def notifications
```

i.e - [discovery](https://github.com/theforeman/foreman_discovery/blob/19f3cdb8cd171749d98150e2d2f4a7304e93ec59/app/views/discovered_hosts/welcome.html.erb#L11) , [foreman_puppet](https://github.com/theforeman/foreman_puppet/blob/866102a4baa24d7b8a5c714ed0f2d1fdb7155885/app/views/foreman_puppet/layouts/application_react.html.erb), [tasks](https://github.com/theforeman/foreman-tasks/blob/49d86fb388cef3945b299c3769f266b6a9b1f5b1/app/views/foreman_tasks/layouts/react.html.erb#L9)
keeping it will be resulted with two `Toasts` components though

tbrisker commented:
nice find @amirfefer! I would suggest in that case replacing the implementation of `notifications` in core so that it only logs a deprecation warning and does not mount the component any longer (since it isn't needed)

Ron-Lavi commented:
> nice find @amirfefer! I would suggest in that case replacing the implementation of `notifications` in core so that it only logs a deprecation warning and does not mount the component any longer (since it isn't needed)

Will do! Thanks @amirfefer @tbrisker

ronlavi2412 committed:
Fixes #32860 - move toast list to app's root

Toasts were triggered only from pages that used the PageLayout component,
but now they will be triggered directly from the root of the app.

ronlavi2412 committed:
Refs #32860 - deprecate notifications method

so toast won't appear twice in pages/plugins where it was re-implemented

ronlavi2412 committed:
Refs #32860 - remove toasts from context

toast doesn't need to be accessed from other components anymore

tbrisker reviewed (approved):
Thanks @laviro !
I love PRs with this kind of change count:
![image](https://user-images.githubusercontent.com/1540531/124625070-26086380-de86-11eb-8ab7-ce98589be7f4.png)
